### PR TITLE
Integration with dandelion

### PR DESCRIPTION
**Pull request by zktuong**
Some initial code that should work with dandelion to convert between the two instances  from https://github.com/zktuong/dandelion/pull/49 onwards.

Still need to work in the extra bits on fully initialized scirpy and dandelion objects.

### CODE REVIEW

File: scirpy/io/_io.py
Comment (grst) [inline scirpy/io/_io.py:531]:
```suggestion
def read_dandelion(dandelion, import_all = False):
```

only class names should start with upper case, not instances thereof

File: scirpy/io/_io.py
Comment (grst) [inline scirpy/io/_io.py:327]:
```suggestion
    if isinstance(path, (str, Path, pd.DataFrame)):
```

(just found out myself that this is possible)

**Review by grst — CHANGES_REQUESTED** [review]
Added some comments, I hope I'm not too nitpicky ;) 

Eventually, we'll need to add tests for that as well, but you can also leave that to me.

**Comment by grst** [comment]
Ah, and you'll need to run the black formatter as well. 
The easiest way to ensure that is to use the pre commit hook - it will format on every git commit. 

To set it up, run the following commands in the root of the repository
```bash
pip install pre-commit
pre-commit install
```

File: scirpy/io/_io.py
Comment (zktuong) [reply scirpy/io/_io.py:365]: it's shortcut to catch NaNs, and also because NaNs can't be lowered.

File: scirpy/io/_io.py
Comment (zktuong) [reply scirpy/io/_io.py:364]: so for this, I tried to keep the loop to be similar to the airr iterator except i can't use `for row in reader:` for a pandas object (it returns the column).

`for index, row in reader.iterrows():` is the alternative which will enable that. not sure if it's possible to do a if-else for different beginnings for a for-loop?

Comment (grst) [reply scirpy/io/_io.py:364]: in that case you can do: 

```python
if isinstance(...): 
   iterator = tmp_path.copy().iterrows()
else:
   iterator = airr.read()
   
for row in iterator: 
   ...
```

Comment (zktuong) [reply scirpy/io/_io.py:364]: nope not working. the iterator with `iterrows()` is returning a tuple `(index, pd.Series)`.

Comment (zktuong) [reply scirpy/io/_io.py:364]: ok i think i found a fix.

```python
if isinstance(tmp_path, pd.DataFrame):
  rows = []
  iterator = tmp_path.copy().iterrows()
  for i in iterator:
    rows.append(i[1])
else:
  rows = airr.read_rearrangement(str(tmp_path))

for row in rows:
  ...
```

Comment (grst) [reply scirpy/io/_io.py:364]: then 
```python
idxiter, iterator = zip(*tmp_path.copy().iterrows())
```

File: scirpy/io/_io.py
Comment (grst) [reply scirpy/io/_io.py]: best use `pd.isnull(x)` instead. 
It also covers the `None` case and is much more explicit.

**Comment by zktuong** [comment]
This is almost ready for review. I'm just trying to solve some bugs with my github actions on `dandelion`'s side holding up further work; seems like the actual tests on this side are working for now, pending updates to #241, but the docs are failing - seems unrelated?.

**Comment by grst** [comment]
The docs complain about something with the type annotations... don't see how this would be related since no type annotations have changed in this PR. 

I'll finally dedicate some time to scirpy next week, but I would like to  finish up #230 first, before I look into #241. 
Is this blocking anything on your side?

**Comment by zktuong** [comment]
Nope all good - take your time.

**Comment by grst** [comment]
This is now merged into #241. I'll keep you posted!

**Pull request closed without merging**
